KEYCLOAK-10659 Proxy authentication support for proxy-mappings

k-tamura opened the pull request:
This is a PR for proxy authentication support for proxy-mappings. I think my PR makes only the minimum necessary changes for the support but it may not be an appropriate solution from the point of view of class design. I have not written test cases yet because test cases may be widely changed responding to how to fix. I would appreciate if you could advise me first (before writing test cases). 

I have tested several proxy-mappings as follows:
```
<property name="proxy-mappings"
 value="[&quot;.*stackexchange\\.com;http://proxyuserid:password@www-proxy.acme.com:8080&quot;,&quot;.*\\.twitter\\.com;NO_PROXY&quot;,&quot;.*;http://fallback:8080/&quot;]"/>
```

See also: [KEYCLOAK-10659](https://issues.redhat.com/browse/KEYCLOAK-10659)

k-tamura committed:
KEYCLOAK-10659 Proxy authentication support for proxy-mappings

hmlnarik commented on services/src/main/java/org/keycloak/connections/httpclient/ProxyMappings.java:
```diff
@@ -181,6 +209,25 @@ private static HttpHost toProxyHost(String proxyUriString) {
+      String[] proxyCredentials = null;
+      if (userInfo != null) {
+        if(userInfo.indexOf(":") > 0){
+          proxyCredentials = userInfo.split(":");
```
```suggestion
          proxyCredentials = userInfo.split(":", 2);
```

k-tamura replied:
Thank you. I will add the `limit` parameter.

hmlnarik commented on services/src/main/java/org/keycloak/connections/httpclient/ProxyMappingsAwareRoutePlanner.java:
```diff
@@ -45,9 +50,16 @@ public ProxyMappingsAwareRoutePlanner(ProxyMappings proxyMappings) {
+    String[] proxyCredentials = proxyMappings.getProxyCredentialsFor(targetHostName);
+    LOG.debugf("Returning proxy=%s for targetHost=%s", proxy, targetHostName);
+    if (proxyCredentials != null && proxyCredentials.length == 2) {
+      NTCredentials ntCreds = new NTCredentials(proxyCredentials[0], proxyCredentials[1], null, null);
```
Please use 4 space indentation
```suggestion
        NTCredentials ntCreds = new NTCredentials(proxyCredentials[0], proxyCredentials[1], null, null);
```

k-tamura replied:
Two spaces and four spaces for indentation are mixed in `ProxyMappings.java` and `ProxyMappingsAwareRoutePlanner.java `. Should I convert all of the former to the latter?

hmlnarik commented on services/src/main/java/org/keycloak/connections/httpclient/ProxyMappings.java:
```diff
@@ -106,6 +109,22 @@ public HttpHost getProxyFor(String hostname) {
+   * @return the proxy credentials associated with the first matching hostname {@link Pattern}
+   * or {@literal null} if none matches.
+   */
+  public String[] getProxyCredentialsFor(String hostname) {
```
This would be a heavily used function. Better use some form of a cache (`ConcurrentMap<String, String[]> hostnameToCreds`) instead of dynamic lookup of credentials with every request.

k-tamura replied:
I see. Please wait a while.

hmlnarik reviewed (changes requested):
Thank you @k-tamura for the PR. Overall looks good.
There is a need to update both tests (see `ProxyMappingsTest`) and documentation (https://github.com/keycloak/keycloak-documentation/blob/master/server_installation/topics/network/outgoing.adoc#proxy-mappings-for-outgoing-http-requests)
Could you please use 4 space indentation though?

k-tamura commented:
> Could you please use 4 space indentation though?

Okay. I will do so.

k-tamura commented:
> There is a need to update both tests (see `ProxyMappingsTest`) and documentation (https://github.com/keycloak/keycloak-documentation/blob/master/server_installation/topics/network/outgoing.adoc#proxy-mappings-for-outgoing-http-requests)

Okay.

k-tamura commented:
Could you review my latest commit? I have updated my code as follows:

- Introduced a proxy-mapping cache.
- Merged two methods, `ProxyMappings#getProxyFor()` and `ProxyMappings#getProxyCredentialsFor()`.
- Added a test case to check if proxy authentication works as expected.
- Changed the password class from `NTCredentials` to `UsernamePasswordCredentials`.

hmlnarik reviewed (changes requested):
Thank you @k-tamura! I am sorry the review took so long.
Looks good, could you please squash the two commits? Then it would be ready to go.

k-tamura commented:
@hmlnarik Thank you for reviewing. I did it. Please merge this.
